File: solutions/python/two_thread_increment.py
import sys
import threading

# @include
N = 1000000
counter = 0


def increment_thread():
    global counter
    for _ in range(N):
        counter = counter + 1

t1 = threading.Thread(target=increment_thread)
t2 = threading.Thread(target=increment_thread)

t1.start()
t2.start()
t1.join()
t2.join()

print(counter)
# @exclude

File: solutions/cpp/two_thread_increment.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <iostream>
#include <string>
#include <thread>

using std::cout;
using std::endl;
using std::stoi;
using std::thread;

// @include
static int counter = 0;

void IncrementThread(int N) {
  for (int i = 0; i < N; ++i) {
    ++counter;
  }
}

void TwoThreadIncrementDriver(int N) {
  thread T1(IncrementThread, N);
  thread T2(IncrementThread, N);
  T1.join();
  T2.join();

  cout << counter << endl;
}
// @exclude

int main(int argc, char* argv[]) {
  int N = argc == 2 ? stoi(argv[1]) : 1000000000;
  TwoThreadIncrementDriver(N);
  return 0;
}
